### ISSUE TITLE
macOS: Neovide hanging on exit

### ISSUE THREAD

**Issue by mizlan** [neovide/neovide#2401]
**Describe the bug**
On macOS, Neovide 0.12.2 hangs on exit. This did not happen on 0.12.1

**To Reproduce**
1. Run `brew install --cask neovide`
2. Open it with `neovide -- --clean`, also happens with `neovide -- -u NORC` (it happens even with no configuration)
3. Notice that editing works fine, but quitting e.g., with <kbd>:q</kbd> or <kbd>ZZ</kbd> causes it to hang.
4. The neovide process pins my CPU to 100%, as seen in htop

**Expected behavior**
Quit with no problem. 

**Screenshots**
<img width="736" alt="Screenshot 2024-03-05 at 1 23 34 PM" src="https://github.com/neovide/neovide/assets/44309097/00ebdf16-bcf2-4087-92a7-b987107c088c">

**Desktop (please complete the following information):**
 - OS: [e.g. macOS Sonoma 14.1.1
 - Neovide Version 0.12.2
 - Neovim Version 

```
NVIM v0.10.0-dev-2424+gbb15fa035-dirty
Build type: Release
LuaJIT 2.1.1707061634
Run "nvim -V1 -v" for more info
```

**Please run `neovide --log` and paste the contents of the `.log` file created in the current directory here:**

As pastebin link: http://sprunge.us/oUHsQf

**Comment by mizlan** [comment]
Oh interesting, I have narrowed it to

```toml
frame = "none"
```

in neovide/config.toml

**Comment by kyoshiro** [comment]
Same here, not in any case but very often when I want to write or close a file Neovide hangs, but still recognizes focus und logs key press:

After hitting ':w' Neovide hangs, last log entry:
```
TRACE [neovide::window::keyboard_manager] Key pressed w ModifiersState(0x0)
TRACE [neovide::channel_utils] UICommand Serial(Keyboard("w"))
TRACE [neovide::window::keyboard_manager] KeyEvent {
    physical_key: Code(
        KeyW,
    ),
    logical_key: Character(
        "w",
    ),
    text: None,
    location: Standard,
    state: Released,
    repeat: false,
    platform_specific: KeyEventExtra {
        text_with_all_modifiers: Some(
            "w",
        ),
        key_without_modifiers: Character(
            "w",
        ),
    },
}
```

But anyway log still fills with aditional keypress registrations and window focus changes:
```
TRACE [neovide::channel_utils] UICommand Parallel(FocusLost)
TRACE [neovide::channel_utils] UICommand Parallel(FocusGained)
TRACE [neovide::window::keyboard_manager] KeyEvent {
    physical_key: Code(
        ControlLeft,
    ),
    logical_key: Named(
        Control,
    ),
    text: None,
    location: Left,
    state: Released,
    repeat: false,
    platform_specific: KeyEventExtra {
        text_with_all_modifiers: None,
        key_without_modifiers: Named(
            Control,
        ),
    },
}
TRACE [neovide::channel_utils] UICommand Parallel(FocusLost)
TRACE [neovide::channel_utils] UICommand Parallel(FocusGained)
TRACE [neovide::channel_utils] UICommand Parallel(FocusLost)
```

**Comment by edithli** [comment]
Same here! 
With config set to be `frame = "none"`, it hangs everytime I exit with ':qa' or ':xa'

**Comment by polachok** [comment]
Try this:

```diff
diff --git a/src/window/mod.rs b/src/window/mod.rs
index 44c6524..460ba4b 100644
--- a/src/window/mod.rs
+++ b/src/window/mod.rs
@@ -18,7 +18,6 @@ use icrate::Foundation::MainThreadMarker;
 use winit::{
     dpi::{PhysicalSize, Size},
     error::EventLoopError,
-    event::Event,
     event_loop::{EventLoop, EventLoopBuilder},
     window::{Icon, Theme, WindowBuilder},
 };
@@ -314,7 +313,8 @@ pub fn main_loop(
     event_loop.run(move |e, window_target| {
         #[cfg(target_os = "macos")]
         menu.ensure_menu_added(&e);
-        if e == Event::LoopExiting {
+
+        if window_target.exiting() {
             return;
         }
```

**Comment by crupest** [comment]
Duplicate of #2087

Please try using `--frame buttonless` temporarily.

**Comment by kyoshiro** [comment]
I'm sorry to say that neither @polachok's supposal nor @crupest mentioned option helped for me.
Neovide still freezes right after writing a file. In some situations I was able to move the cursor once before freeze.

**Comment by kyoshiro** [comment]
Hi @crupest, I tested with option `frame = "buttonless"` in config file and I can confirm the buttons and title bar disappeared. I then opened some files and saving one of those files freezed neovide. Maybe worth to  mention, that the status line shows these symbols when it came to freeze: 
![image](https://github.com/neovide/neovide/assets/108440/1745e1a5-ad07-44fa-a05f-ef9bcdbbeeda)

**Comment by crupest** [comment]
@kyoshiro Well, really thank you for your testing. It's just a little weird because I can't reproduce the bug in my situation. Only `--frame none` freezes on my machine and `--frame buttonless` does not. If you are willing, maybe you can try another GUI app of neovim and see if it freezes in your case. If it does not, then it is definitely a bug of neovide.

**Comment by crupest** [comment]
@kyoshiro And the bug should only trigger when you quit neovim with `:q` or other quit command. But based on what you said, just `:w` triggers the hanging.

**Comment by kyoshiro** [comment]
Hi @crupest yes, I will give it some more testing 2mrw. In my experience the hanging happens on :w :x :wq or :q. This does not occur when using nvim natively in terminal or tmux and I also did not recognize the hanging in neovide on Linux. So, it's kind of detective work on Mac right now but I like to figure it out.

**Comment by fredizzimo** [comment]
From the logs it looks like the event loop does not exit for some reason, even `window_target.exit();` is repeatedly called, there's a repeated log of "Saved Window Settings", which implies that it's called, since it's inside the same if block. 

And I think we have debugged this already once in Discord, but I can't find it right now, and commenting out `save_window_size` should "fix" the issue.

The cause is quite probably this, from https://rust-windowing.github.io/winit/winit/platform/macos/index.html
> A lot of functionality expects the application to be ready before you start doing anything; this includes creating windows, fetching monitors, drawing, and so on, see issues [#2238](https://github.com/rust-windowing/winit/issues/2238), [#2051](https://github.com/rust-windowing/winit/issues/2051) and [#2087](https://github.com/rust-windowing/winit/issues/2087).

> If you encounter problems, you should try doing your initialization inside Event::Resumed.

I think @falcucci, might look into fixing that at some point.

**Comment by fredizzimo** [comment]
I found it, starting from here:
https://discordapp.com/channels/896132176091955261/896263894086066216/1194132583701811240

There are some interesting other observations, there like https://github.com/rust-windowing/winit/blob/43f29f048180b24479b5406118f7e04f52befa30/src/platform_impl/macos/window.rs#L905, which modifies the windows attributes in order to determine if it's zoomed or not. So, the resolution might be something different too, maybe to make sure that we call `save_window_size` only once.

**Comment by kyoshiro** [comment]
Hi @fredizzimo, great. I commented the save_window_size in window/mod.rs and looks quite good at first glance. Giving it more testing over the day. I also reverted changes made before to have an atomic change here.
My preferred setting for frame is: `frame = "transparent"` on MacOS.

**Comment by kyoshiro** [comment]
Unfortunately the problem still exists. It's kind of curious, because it seems to happen more often after time and is reproducable more instantly.

**Comment by crupest** [comment]
@kyoshiro That makes much more sense then, though it's very strange only neovide triggers the problem. Maybe @fredizzimo has more insights.

**Comment by kyoshiro** [comment]
@crupest @fredizzimo I spent some additional time in testing and now I can confirm, it's definitely neovide in combination with terraform fmt. I was able to reproduce the behavior on Linux (Arch), too. So maybe it's time to split this up as I do not know if it's related to the initial opening question of this issue.

**Comment by fredizzimo** [comment]
Yes, I think there are at least two different causes here. The original issue was reported to happen even with `--clean`

**Comment by mediapathic** [comment]
Just an additional datapoint: I am also seeing the freeze on `:q` with `--frame none`, but not with `--frame buttonless`, and I have never even heard of a terraform file :) . (Mac, latest stable release)

**Comment by falcucci** [comment]
this is fixed after https://github.com/neovim/neovim/pull/28157 got merged.
